docs/dev/guidelines/server.rst

```diff
@@ -350,3 +350,240 @@ Please read `the AssertJ documentation <https://assertj.github.io/doc/#assertj-c
+        return switch (language) {
+            case TranslationService.LANGUAGE_GERMAN -> translateToGerman(word);
+            // TODO: Add a case for the French language
+            default -> throw new IllegalStateException("Illegal language requested: " + language);
+        };
+    }
+
+    public String runService(String serviceName, String parameter) {
+        String result = null;
+        if (serviceName.equals(TranslationService.SERVICE_NAME_TRANSLATION_GERMAN)) {
+            result = translate(parameter, TranslationService.LANGUAGE_GERMAN);
+        }
+        // TODO: Add a case for the French language
+        else {
+            System.out.println("Can't offer service " + serviceName);
+        }
+        return result;
+    }
+
+There are two separate tests, one testing the translation and the other one testing the runService method.
+The test for runService must not assume that the translate method is already implemented correctly. A possible solution for this problem could look like this:
+
+.. code-block:: java
+
+    @Test
+    public void testRunServiceInTranslationServer() {
+        TranslationServer testTranslationServer = new TranslationServer() {
+            public String translate(String word, String language) {
+                return word + ":" + language;
+            }
+        }
+        String expected = "Dog:French";
+        String actual = testTranslationServer.runService("Dog", "French");
+        if(!expected.equals(actual)) fail("Descriptive fail message");
+    }
+
+This test correctly tests if the student added the case for the French language and called the appropriate method with the appropriate parameters. Because the translation method was overridden,
+it doesn't matter whether the student has already completed the previous task or not.
+(If you do that, you should have some way to deal with students that make the class or method final, either via problem statement or test, otherwise students get compilation errors in the test code)
+
+
+Make use of JUnit 5 Features:
+https://junit.org/junit5/docs/current/user-guide/#writing-tests
+https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html
+
+General Testing best practices are found on https://phauer.com/2019/modern-best-practices-testing-java/
+
+Also check out https://www.baeldung.com/spring-tests
+
+And https://howtodoinjava.com/best-practices/unit-testing-best-practices-junit-reference-guide/
+
+
+23. Avoid using @MockBean
+=========================
+
+Do not use the ``@SpyBean`` or ``@MockBean`` annotation unless absolutely necessary, or possibly in an abstract Superclass. If you want to see more in detail why, take a look `here. <https://www.baeldung.com/spring-tests>`_
+Basically every time @MockBean appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.
```

Comment:
You could unify the use of ``@MockBean`` and @Mockbean

Reply:
```suggestion
Basically every time ``@MockBean`` appears in a class, the ApplicationContext cache gets marked as dirty, hence the runner will clean the cache after the test-class is done and restarts the application context.
```
like this you mean?